request.go

```diff
@@ -1,196 +1,135 @@
 
-	prof := log.StartProfiler("serve %s", r.method)
-	arg, err := r.getArg(receiver, handler)
+	prof := r.log.StartProfiler("serve %s", r.Name())
+	arg := r.Arg()
 
-	go func() {
-		r.LogInvocation(log, err, arg)
-		if err != nil {
-			r.err = wrapError(wrapErrorFunc, err)
-		} else {
-			res, err := handler.Handler(r.ctx, arg)
-			r.err = wrapError(wrapErrorFunc, err)
-			r.res = res
-		}
-		prof.Stop()
-		r.LogCompletion(log, err)
-		r.Reply(transmitter, log)
-	}()
+	r.LogInvocation(nil)
+	res, err := handler.Handler(r.ctx, arg)
+	prof.Stop()
+	r.LogCompletion(res, err)
+
+	r.Reply(transmitter, res, wrapError(wrapErrorFunc, err))
 }
 
 type notifyRequest struct {
 	requestImpl
 }
 
-func newNotifyRequest() *notifyRequest {
+func newNotifyRequest(rpc RPCMessage, log LogInterface) *notifyRequest {
 	ctx, cancel := context.WithCancel(context.Background())
-	r := &notifyRequest{
+	return &notifyRequest{
 		requestImpl: requestImpl{
-			message: message{
-				remainingFields: 2,
-			},
+			RPCMessage: rpc,
 			ctx:        ctx,
 			cancelFunc: cancel,
+			log:        log,
 		},
 	}
-	r.decodeSlots = []interface{}{
-		&r.method,
-	}
-	return r
-}
-
-func (r *notifyRequest) LogInvocation(log LogInterface, err error, arg interface{}) {
-	log.ServerNotifyCall(r.method, err, arg)
 }
 
-func (r *notifyRequest) LogCompletion(log LogInterface, err error) {
-	log.ServerNotifyComplete(r.method, err)
+func (r *notifyRequest) LogInvocation(err error) {
+	r.log.ServerNotifyCall(r.Name(), err, r.Arg())
 }
 
-func (r *notifyRequest) Serve(receiver byteReadingDecoder, transmitter encoder, handler *ServeHandlerDescription, wrapErrorFunc WrapErrorFunc, log LogInterface) {
-
-	prof := log.StartProfiler("serve-notify %s", r.method)
-	arg, err := r.getArg(receiver, handler)
-
-	go func() {
-		r.LogInvocation(log, err, arg)
-		if err == nil {
-			_, err = handler.Handler(r.ctx, arg)
-		}
-		prof.Stop()
-		r.LogCompletion(log, err)
-	}()
+func (r *notifyRequest) LogCompletion(_ interface{}, err error) {
+	r.log.ServerNotifyComplete(r.Name(), err)
 }
 
-type cancelRequest struct {
-	requestImpl
-}
+func (r *notifyRequest) Serve(transmitter encoder, handler *ServeHandlerDescription, wrapErrorFunc WrapErrorFunc) {
 
-func newCancelRequest() *cancelRequest {
-	r := &cancelRequest{
-		requestImpl: requestImpl{
-			message: message{
-				remainingFields: 2,
-			},
-			ctx: context.Background(),
-		},
-	}
-	r.decodeSlots = []interface{}{
-		&r.seqno,
-		&r.method,
-	}
-	return r
-}
+	prof := r.log.StartProfiler("serve-notify %s", r.Name())
+	arg := r.Arg()
 
-func (r *cancelRequest) LogInvocation(log LogInterface, err error, arg interface{}) {
-	log.ServerCancelCall(r.seqno, r.method)
+	r.LogInvocation(nil)
+	_, err := handler.Handler(r.ctx, arg)
+	prof.Stop()
+	r.LogCompletion(nil, err)
 }
 
-func newRequest(methodType MethodType) request {
-	switch methodType {
+func newRequest(rpc RPCMessage, log LogInterface) request {
```

Comment:
I don't think you need this function anymore, either -- all the callers already know the type, so they can just call the right function directly.